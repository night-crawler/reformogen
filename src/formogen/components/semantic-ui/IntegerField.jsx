import _ from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import { Form } from 'semantic-ui-react';
import { errorsType, layoutOptsType } from '../fieldPropTypes';
import Label from './Label';
import { MessageList } from './MiscComponents';


IntegerField.propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    min_value: PropTypes.number,
    max_value: PropTypes.number,
    decimal_places: PropTypes.number,
    help_text: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    errors: errorsType,

    required: PropTypes.bool,
    editable: PropTypes.bool,

    helpTextOnHover: PropTypes.bool,
    layoutOpts: layoutOptsType,

    updateProps: PropTypes.func,
    onChange: PropTypes.func,
};
export default function IntegerField(props) {
    let _props = {
        name: props.name,
        value: props.value,
        placeholder: props.placeholder,
        onChange: props.onChange,
    };

    let input_options = {
        type: 'number',
        min: props.min_value || undefined,
        max: props.max_value || undefined,
    };

    switch(props.type) {
        case 'PositiveSmallIntegerField':
            if (input_options.min === undefined) {input_options.min = 0;}
            if (input_options.max === undefined) {input_options.max = 32767;}
            break;

        case 'PositiveIntegerField':
            if (input_options.min === undefined) {input_options.min = 0;}
            if (input_options.max === undefined) {input_options.max = 2147483647;}
            break;

        // https://docs.djangoproject.com/en/1.11/ref/models/fields/#smallintegerfield
        case 'SmallIntegerField':
            if (input_options.min === undefined) {input_options.min = -32768;}
            if (input_options.max === undefined) {input_options.max = 32767;}
            break;

        case 'IntegerField':
            if (input_options.min === undefined) {input_options.min = -2147483648;}
            if (input_options.max === undefined) {input_options.max = 2147483647;}
            break;

        case 'FloatField':
            input_options.step = 0.1;
            break;

        case 'DecimalField':
            input_options.step = 0.1;
            if (props.decimal_places) {
                if (props.decimal_places >= 2) {
                    input_options.step = 1 / Math.pow(10, (props.decimal_places / 2));
                }
            }
            break;

        default:
            break;
    }

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field
            required={ props.required }
            disabled={ !props.editable }
            width={ props.layoutOpts.width }
            error={ !_.isEmpty(props.errors) }
        >
            <Label { ...props } />
            <Form.Input { ..._props } { ...input_options } />
            { !props.helpTextOnHover ? <span className="help-text">{ props.help_text }</span> : ''}
            <MessageList messages={ props.errors } />
        </Form.Field>
    );
}
