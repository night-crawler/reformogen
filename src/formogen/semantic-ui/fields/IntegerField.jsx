import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import { isEmpty } from 'lodash';

import { errorsType, displayOptionsType } from '../../fieldPropTypes';
import { FieldLabel } from '../FieldLabel';
import { ErrorsList } from '../ErrorsList';


IntegerField.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  min_value: PropTypes.number,
  max_value: PropTypes.number,
  decimal_places: PropTypes.number,
  help_text: PropTypes.string,
  placeholder: PropTypes.string,
  errors: errorsType,

  required: PropTypes.bool,
  editable: PropTypes.bool,

  helpTextOnHover: PropTypes.bool,
  displayOptions: displayOptionsType,

  onChange: PropTypes.func,
};
export function IntegerField(props) {
  let inputOptions = {
    type: 'number',
    min: props.min_value || undefined,
    max: props.max_value || undefined,
  };

  switch(props.type) {
    case 'PositiveSmallIntegerField':
      if (inputOptions.min === undefined) {inputOptions.min = 0;}
      if (inputOptions.max === undefined) {inputOptions.max = 32767;}
      break;

    case 'PositiveIntegerField':
      if (inputOptions.min === undefined) {inputOptions.min = 0;}
      if (inputOptions.max === undefined) {inputOptions.max = 2147483647;}
      break;

      // https://docs.djangoproject.com/en/1.11/ref/models/fields/#smallintegerfield
    case 'SmallIntegerField':
      if (inputOptions.min === undefined) {inputOptions.min = -32768;}
      if (inputOptions.max === undefined) {inputOptions.max = 32767;}
      break;

    case 'IntegerField':
      if (inputOptions.min === undefined) {inputOptions.min = -2147483648;}
      if (inputOptions.max === undefined) {inputOptions.max = 2147483647;}
      break;

    case 'FloatField':
      inputOptions.step = 0.1;
      break;

    case 'DecimalField':
      inputOptions.step = 0.1;
      if (props.decimal_places) {
        if (props.decimal_places >= 2) {
          inputOptions.step = 1 / Math.pow(10, (props.decimal_places / 2));
        }
      }
      break;

    default:
      break;
  }

  return (
    <Form.Field
      required={ props.required }
      disabled={ !props.editable }
      width={ props.displayOptions.width }
      error={ !isEmpty(props.errors) }
    >
      <FieldLabel { ...props } />
      <Form.Input 
        name={ props.name }
        value={ props.value }
        placeholder={ props.placeholder }
        onChange={ props.onChange }
        { ...inputOptions } 
      />
      { !props.helpTextOnHover
        ? <span className='help-text'>{ props.help_text }</span>
        : ''
      }
      <ErrorsList messages={ props.errors } />
    </Form.Field>
  );
}
