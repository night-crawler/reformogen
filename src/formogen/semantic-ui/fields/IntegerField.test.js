import { IntegerField } from './IntegerField';


it('<IntegerField />', () => {
  const wrapper = shallow(
    <IntegerField
      type='IntegerField'
      name='IntegerField: name'
      verbose_name='IntegerField: verbose_name'
      help_text='IntegerField: help_text'
      layoutOpts={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
