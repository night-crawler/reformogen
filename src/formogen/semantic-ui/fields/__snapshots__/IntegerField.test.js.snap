// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IntegerField 1`] = `
<FormField
  disabled={true}
  error={false}
  width={1}
>
  <Label
    help_text="IntegerField: help_text"
    layoutOpts={
      Object {
        "width": 1,
      }
    }
    name="IntegerField: name"
    type="IntegerField"
    verbose_name="IntegerField: verbose_name"
  />
  <FormInput
    as={[Function]}
    control={[Function]}
    max={2147483647}
    min={-2147483648}
    name="IntegerField: name"
    type="number"
  />
  <span
    className="help-text"
  >
    IntegerField: help_text
  </span>
  <ErrorsList
    color="red"
    header="Errors"
  />
</FormField>
`;
